chainer/functions/activation/hard_sigmoid.py

```diff
@@ -19,30 +19,49 @@ def check_type_forward(self, in_types):
     def forward_cpu(self, inputs):
         x = inputs[0]
         y = numpy.clip(x * 0.2 + 0.5, 0.0, 1.0)
+        self.retain_inputs((0,))
         return utils.force_array(y, x.dtype),
 
     def forward_gpu(self, inputs):
         x = inputs[0]
+        self.retain_inputs((0,))
         return cuda.elementwise(
             'T x', 'T y',
             'y = min(1.0, max(0.0, x * 0.2 + 0.5))',
             'hard_sigmoid_fwd'
         )(x),
 
-    def backward_cpu(self, inputs, grads):
-        x = inputs[0]
-        g = grads[0]
-        gx = ((-2.5 < x) & (x < 2.5)) * g * 0.2
-        return utils.force_array(gx, x.dtype),
+    def backward(self, indexes, grad_outputs):
+        x, = self.get_retained_inputs()
+        return HardSigmoidGrad(x.data).apply(grad_outputs)
 
-    def backward_gpu(self, inputs, grads):
-        x = inputs[0]
-        g = grads[0]
+
+class HardSigmoidGrad(function_node.FunctionNode):
+
+    """Hard-sigmoid gradient function."""
+
+    def __init__(self, x):
+        self.x = x
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+        type_check.expect(in_types[0].dtype.kind == 'f')
```

Comment:
Check `in_types[0].dtype == self.x.dtype`

Reply:
Added this type check.